### PR TITLE
[IMPROVE] Send files over rest api

### PR DESCRIPTION
**Pull request by ggazzo**

### CODE REVIEW

File: app/utils/client/lib/RestApiClient.js [hunk 1]
```diff
@@ -68,15 +68,32 @@ export const APIClient = {
 		});
 	},
 
-	_jqueryFormDataCall(endpoint, params, formData) {
+	_jqueryFormDataCall(endpoint, params, formData, { progress = () => {}, error = () => {} } = {}, abort = () => {}) {
```
Comment (sampaiodiego) [inline app/utils/client/lib/RestApiClient.js]:
```suggestion
	_jqueryFormDataCall(endpoint, params, formData, { progress = () => {}, error = () => {} } = {}) {
```

File: app/utils/client/lib/RestApiClient.js [hunk 2]
```diff
@@ -68,15 +68,32 @@ export const APIClient = {
...
+
+					xhr.upload.addEventListener('progress', function(evt) {
+						if (evt.lengthComputable) {
+							const percentComplete = evt.loaded / evt.total;
+							progress(percentComplete * 100);
+						}
+					}, false);
+
+					xhr.upload.addEventListener('error', error, false);
+					xhr.upload.addEventListener('abort', abort, false);
```
Comment (sampaiodiego) [inline app/utils/client/lib/RestApiClient.js]:
```suggestion
```

**Comment by geekgonecrazy** [comment]
I think we still need to look at chunking or multi-part.  Most proxies are going to have 20MB limit for payload. Even on cloud we limit to 20MB payload I believe.  

I think only reason this hasn't been an issue before is because mobile has been the only one really sending files this way until now.

**Comment by recklessnl** [comment]
@geekgonecrazy Thanks for looking into this issue. Especially on local networks, the transfer speed should roughly match the general transfer speed of the network, and that currently isn't even close. 

On my network, for example, I can transfer files to the same Rocket.Chat server a 125MB/s, maximizing the gigabit connection. I cannot get anywhere near this with Rocket Chat file transfers, as explained in #16542. 

Even a 20MB limit would probably already help tremendously, but ideally this would be a tuneable setting in the admin console, as purely local networks need not to worry about these proxy limits.

**Pull request merged**